File: web/src/services/api/notifications.ts
import { request } from './client';

export type NotificationItem = {
  id: string;
  title: string;
  body: string;
  type: string;
  grade: string;
  section: string;
  read: boolean;
  created_at: string;
};

const STORAGE_KEY = 'edusync-notifications';

function readLocalNotifications(grade: string, section: string): NotificationItem[] {
  try {
    const raw = window.localStorage.getItem(STORAGE_KEY);
    if (!raw) return [];
    const items = JSON.parse(raw) as NotificationItem[];
    return items.filter(item => item.grade === grade && item.section.toUpperCase() === section.toUpperCase());
  } catch {
    return [];
  }
}

function writeLocalNotifications(items: NotificationItem[]) {
  window.localStorage.setItem(STORAGE_KEY, JSON.stringify(items));
}

export function addLocalTimetableNotification(grade: string, section: string) {
  const existing = readLocalNotifications(grade, section);
  const allRaw = window.localStorage.getItem(STORAGE_KEY);
  const all = allRaw ? (JSON.parse(allRaw) as NotificationItem[]) : [];
  const notification: NotificationItem = {
    id: `local-${Date.now()}`,
    title: 'Timetable Updated',
    body: `Your Grade ${grade} • Section ${section} timetable has been updated. Check your schedule for the latest periods.`,
    type: 'timetable',
    grade,
    section: section.toUpperCase(),
    read: false,
    created_at: new Date().toISOString(),
  };
  writeLocalNotifications([notification, ...all.filter(item => item.id !== notification.id)]);
  return [notification, ...existing];
}

export async function getNotifications(grade = '12', section = 'A') {
  try {
    return await request<NotificationItem[]>(`/notifications?grade=${encodeURIComponent(grade)}&section=${encodeURIComponent(section)}`);
  } catch {
    return readLocalNotifications(grade, section);
  }
}

export async function markNotificationRead(notificationId: string) {
  try {
    await request<{ read: boolean }>(`/notifications/${encodeURIComponent(notificationId)}/read`, { method: 'PATCH' });
  } catch {
    const raw = window.localStorage.getItem(STORAGE_KEY);
    if (!raw) return;
    const items = JSON.parse(raw) as NotificationItem[];
    writeLocalNotifications(items.map(item => item.id === notificationId ? { ...item, read: true } : item));
  }
}

File: backend/app/services/notifications.py
from dataclasses import asdict, dataclass
from datetime import datetime, timezone
from uuid import uuid4


@dataclass
class Notification:
    id: str
    title: str
    body: str
    type: str
    grade: str
    section: str
    read: bool
    created_at: str


_store: list[Notification] = []


def create_timetable_notification(grade: str, section: str) -> Notification:
    notification = Notification(
        id=str(uuid4()),
        title="Timetable Updated",
        body=f"Your Grade {grade} • Section {section} timetable has been updated. Check your schedule for the latest periods.",
        type="timetable",
        grade=grade,
        section=section.upper(),
        read=False,
        created_at=datetime.now(timezone.utc).isoformat(),
    )
    _store.insert(0, notification)
    return notification


def list_notifications(grade: str | None = None, section: str | None = None) -> list[dict]:
    items = _store
    if grade and section:
        items = [item for item in _store if item.grade == grade and item.section == section.upper()]
    return [asdict(item) for item in items]


def mark_read(notification_id: str) -> bool:
    for item in _store:
        if item.id == notification_id:
            item.read = True
            return True
    return False
